src/System.Collections.Concurrent/src/System/Collections/Concurrent/ConcurrentDictionary.cs

```diff
@@ -905,7 +904,7 @@ private static void ThrowKeyNullException()
 
 
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        private static void ThrowIfInvalidObjectValue(object value)
+        private static void ThrowIfInvalidObjectValue(object? value)
```

Comment:
How does `if (key == null) ThrowKeyNullException();` above work? TKey can be struct, right? Is it boxed for the sake of comparison?

Reply:
If the C# compiler can see that the TKey is a struct (other than a nullable value type), then either it'll simply eliminate the comparison with a warning saying that it would always be false (in the case where you use the explicit type) or it'll outright fail to compile (in the case where you use a struct constraint on the generic).  If it can't tell, it'll generate a box instruction in the IL, but then the JIT will see that it'll never be null, and will similarly eliminate the boxing and the comparison entirely.